### PR TITLE
libroach: use FlushWAL instead of SyncWAL

### LINKED ISSUE #22443: engine: Try rocksdb `manual_wal_flush`/`FlushWAL`
RocksDB has independent options for flushing (the `write()` system call) and syncing (`fdatasync()`) data. We sync roughly half of our writes, but we still flush all of the unsynced writes. We might be able to improve performance if we set `manual_wal_flush` in options and used `FlushWAL` instead of `SyncWAL` in the sync loop. This would allow small non-synced writes to be fully batched with the system calls of the next synced write.

### PR DESCRIPTION
**Pull request by petermattis**
Tell RocksDB that we will manually flush the wal via the
`manual_wal_flush` option. Replace `SyncWAL()` with `FlushWAL(true)`. We
"sync" approximately half of writes. The effect of this change is to
avoid calling `write(2)` on the writes which we are not syncing.

Fixes #22443

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
@bdarnell Initial testing on my laptop shows that this does provide a performance boost. I'll run some longer tests on a real cluster today to provide accurate before and after numbers.

**Comment by bdarnell** [comment]
:lgtm: assuming performance looks good. 

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22457#-:-L4l2Ekci5qcAFFX7h__:b-hdzict)*

**Comment by petermattis** [comment]
On a 6-node cluster running `kv --read-percent=0`, the performance benefit currently looks to be in the 2-5% range. Not nothing, but not huge. 

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22457#-:-L4l5tHt_cKo1w1D3gbZ:b-a7mkae)*

**Comment by petermattis** [comment]
Rebased and will merge on green, but I'm intentionally not cherry-picking this change.

**Comment by petermattis** [comment]
There appears to be a concurrency bug in RocksDB when using `Options::manual_wal_flush` without `Options::concurrent_prepares`. See https://github.com/facebook/rocksdb/issues/3599. The following diff fixes the problem:

```
diff --git a/db/db_impl_write.cc b/db/db_impl_write.cc
index 67ebc08f..18bddf3d 100644
--- a/db/db_impl_write.cc
+++ b/db/db_impl_write.cc
@@ -240,8 +240,14 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
     if (!concurrent_prepare_) {
       if (status.ok() && !write_options.disableWAL) {
         PERF_TIMER_GUARD(write_wal_time);
+        if (manual_wal_flush_) {
+          log_write_mutex_.Lock();
+        }
         status = WriteToWAL(write_group, log_writer, log_used, need_log_sync,
                             need_log_dir_sync, last_sequence + 1);
+        if (manual_wal_flush_) {
+          log_write_mutex_.Unlock();
+        }
       }
     } else {
       if (status.ok() && !write_options.disableWAL) {
```

**Comment by nvanbenschoten** [comment]
I rebased this on top of https://github.com/cockroachdb/cockroach/pull/26810 and ran some experiments on it. I wasn't able to see a noticeable speedup when running `kv --read-percent=0 --splits=10` against a 6 node cluster with `--concurrency` set to `16`, regardless of whether the `nobarrier` option was enabled or not. However, when I bumped the concurrency up to `320` I was able to see about a 4% throughput improvement, which lines up with what you saw before. In conjunction with your previous testing, I think that demonstrates that this is a worthwhile change.

**Comment by petermattis** [comment]
@nvanbenschoten Ack. I've rebased as well and removed the `DO NOT MERGE` message.

**Review by nvb — APPROVED** [review]

**Comment by petermattis** [comment]
Finally tracked down why my build is failing and discovered:

```
tocommit(46) is out of range [lastIndex(38)]. Was the raft log corrupted, truncated, or lost?
```

Need to check to see if the upstream fix for the problem we experienced with `FlushWAL` is present.

**Comment by petermattis** [comment]
Hmm, https://github.com/facebook/rocksdb/pull/3656 claims to have fixed the upstream issue, but perhaps it didn't. The failure above came from `make acceptance TESTS='TestDecommission/runMode=local'.

**Comment by petermattis** [comment]
I'm not sure why, but RocksDB v5.13.4 does not include https://github.com/facebook/rocksdb/pull/3656/commits/c3f1693a79421c843ab1601b4579b4ed150a9a8c. That commit merged on Mar 26 while 5.13.4 was released on June 18th. Regardless of the reason, this PR can't merge until we have a RocksDB release with that commit.

**Comment by petermattis** [comment]
Ah, RocksDB uses long-lived release branches. The v5.13 release branch was created on Mar 20 and they seem to only cherry-pick changes into the release branch after creation. It looks like  facebook/rocksdb@c3f1693 was never cherry-picked. We'll either need to cherry-pick it ourselves, or wait for v5.14.

**Comment by petermattis** [comment]
@benesch I cherry-picked facebook/rocksdb@c3f1693 into our RocksDB fork. Can you do a quick sanity check that I did this correctly?

**Comment by benesch** [comment]
LGTM, but please push that commit to the tip of the crl-release-5.13 branch in cockroachdb/rocksdb!

**Comment by petermattis** [comment]
@benesch Done.

**Comment by petermattis** [comment]
bors r=nvanbenschoten,benesch

**Comment by petermattis** [comment]
Acceptance test CI was canceled?

bors r=nvanbenschoten,benesch

**Comment by benesch** [comment]
Yeah, I've noticed that happening lately. It's quite frustrating.

**Pull request merged**
